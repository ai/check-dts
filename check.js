let { dirname, basename, join, relative } = require('path')
let { promisify } = require('util')
let lineColumn = require('line-column')
let globby = require('globby')
let chalk = require('chalk')
let ora = require('ora')
let fs = require('fs')

let readFile = promisify(fs.readFile)

const TS_DIR = dirname(require.resolve('typescript'))
const TS_CONFIG_FILE = './tsconfig.json'
const WORKER = join(__dirname, 'worker.js')
const PREFIX = '// THROWS '

let Worker, createProgram
try {
  Worker = require('worker_threads').Worker
} catch {
  createProgram = require('./create-program')
}

let r = chalk.red
let b = chalk.bold
let g = chalk.green

function checkFiles (files, compilerOptions) {
  if (Worker) {
    return new Promise((resolve, reject) => {
      let worker = new Worker(WORKER, {
        workerData: { files, compilerOptions }
      })
      worker.on('message', resolve)
      worker.on('error', reject)
    })
  } else {
    return createProgram(files, compilerOptions)
  }
}

function getText (error) {
  if (typeof error.messageText === 'object') {
    return error.messageText.messageText
  } else {
    return error.messageText
  }
}

function formatName (cwd, file) {
  return chalk.gray(
    relative(cwd, file).replace(basename(file), i => chalk.white.bold(i))
  )
}

async function parseTest (files) {
  let expects = []
  await Promise.all(
    files.map(async fileName => {
      let source = (await readFile(fileName)).toString()
      let prev, pos
      let lines = lineColumn(source)
      while (true) {
        pos = source.indexOf(PREFIX, prev + 1)
        if (pos === -1) break
        let newline = source.indexOf('\n', pos)
        let pattern = source.slice(pos + PREFIX.length, newline)
        let { line, col } = lines.fromIndex(pos)
        expects.push({ fileName, line: line + 1, col, pattern })
        prev = pos
      }
    })
  )
  return expects
}

module.exports = async function check (stdout, cwd, print) {
  let spinner = ora({ stream: stdout }).start('Check types')

  let compilerOptions = null
  if (fs.existsSync(TS_CONFIG_FILE)) {
    let tsConfig = JSON.parse(await readFile(TS_CONFIG_FILE))
    compilerOptions = tsConfig.compilerOptions || null
  }

  let opts = { cwd, ignore: ['node_modules'], gitignore: true, absolute: true }
  let all = await globby('**/*.{js,ts,jsx,tsx}', opts)

  if (!all.some(i => /\.tsx?$/.test(i))) {
    let err = new Error(
      'TypeScript files was not found. ' +
      'Create .d.ts files and test/types.ts and test/errors.ts.'
    )
    err.own = true
    throw err
  }

  let typeTests = all.filter(i => /\.?(errors|types)\.tsx?/.test(i))
  let failTests = all.filter(i => /\.?errors\.tsx?$/.test(i))

  all.push(join(TS_DIR, 'lib.dom.d.ts'))
  all.push(join(TS_DIR, 'lib.es2018.d.ts'))

  let expects = await parseTest(failTests)
  let errors = await checkFiles(all, compilerOptions)

  let bad = { }
  function push (file, message) {
    if (!bad[file]) bad[file] = []
    bad[file].push(message)
  }

  for (let i of errors) {
    let { line, col } = lineColumn(i.file.text).fromIndex(i.start)
    let expect = expects.find(j => {
      return i.file.fileName === j.fileName && line === j.line && !j.used
    })
    let pos = r(`${ line }:${ col }:`)
    let text = getText(i)
    if (!failTests.includes(i.file.fileName)) {
      push(
        i.file.fileName,
        '  ' + b(pos + ' Type error ' + chalk.gray(`TS${ i.code }`) + '\n') +
        '  ' + r(text)
      )
    } else if (!expect) {
      push(
        i.file.fileName,
        '  ' + b(pos + ' Unexpected error\n') +
        '  ' + r(text)
      )
    } else {
      expect.used = true
      if (!text.includes(expect.pattern)) {
        push(
          i.file.fileName,
          '  ' + b(pos + ' Wrong error\n') +
          '  Expected: ' + g(expect.pattern) + '\n' +
          '  Got: ' + r(text)
        )
      }
    }
  }
  let unused = expects.filter(i => !i.used)
  for (let i of unused) {
    push(
      i.fileName,
      '  ' + b(r(`${ i.line }:${ i.col }:`) + ' Error was not found\n') +
      '  ' + r(i.pattern)
    )
  }

  let failed = Object.keys(bad).length > 0
  if (failed) {
    spinner.fail()
    for (let file in bad) {
      print(r('✖ ') + formatName(cwd, file) + '\n')
      let messages = (bad[file] || []).sort((msg1, msg2) => {
        let line1 = parseInt(msg1.match(/(\d+):/)[1])
        let line2 = parseInt(msg2.match(/(\d+):/)[1])
        return line1 - line2
      })
      for (let i of messages) {
        print(i + '\n')
      }
    }
    return false
  } else {
    spinner.succeed()
    for (let i of typeTests) {
      print(chalk.green('✔ ') + formatName(cwd, i))
    }
    return true
  }
}
