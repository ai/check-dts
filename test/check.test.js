import { fileURLToPath } from 'url'
import { join, dirname } from 'path'
import { jest } from '@jest/globals'

import './fixtures/ci.js'
import { check } from '../check.js'

const ROOT = dirname(fileURLToPath(import.meta.url))

async function run(fixture, args) {
  let out = { text: '' }
  out.write = symbols => {
    out.text += symbols
  }
  out.print = (...lines) => {
    out.write(lines.join('\n') + '\n')
  }
  let cwd = join(ROOT, 'fixtures', fixture)
  let exitCode = await check(out, cwd, out.print, args)
  return [exitCode, out.text]
}

async function good(fixture, args) {
  let [exitCode, out] = await run(fixture, args)
  expect(exitCode).toBe(true)
  return out
}

async function bad(fixture, args) {
  let [exitCode, out] = await run(fixture, args)
  expect(exitCode).toBe(false)
  return out
}

jest.setTimeout(20000)

it('supports simple cases', async () => {
  expect(await good('simple')).toMatchSnapshot()
})

it('checks positive tests', async () => {
  expect(await bad('positive')).toMatchSnapshot()
})

it('checks negative tests', async () => {
  expect(await bad('negative')).toMatchSnapshot()
})

it('checks both tests', async () => {
  expect(await bad('both')).toMatchSnapshot()
})

it('checks mixed tests', async () => {
  expect(await good('mixed')).toMatchSnapshot()
})

it('loads custom tsconfig.json', async () => {
  expect(await good('tsconfig')).toMatchSnapshot()
})

it('loads custom tsconfig.json with extends property', async () => {
  expect(await good('tsconfig-with-extends')).toMatchSnapshot()
})

it('accepts files', async () => {
  expect(await good('negative', 'b.*')).toMatchSnapshot()
})

it('warns about empty project', async () => {
  let error
  try {
    await bad('empty')
  } catch (e) {
    error = e
  }
  expect(error.message).toContain('TypeScript files were not found')
  expect(error.own).toBe(true)
})
