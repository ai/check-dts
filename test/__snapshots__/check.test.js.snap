// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts files 1`] = `
"[33m-[39m Check types
[32mâœ”[39m Check types
"
`;

exports[`checks both tests 1`] = `
"[33m-[39m Check types
[31mâœ–[39m Check types

[31mâœ– [90mtest/[31ma.types.ts[90m[31m:3:10:[39m[1m Type error [90mTS2345[39m
[22m  [31mArgument of type 'string' is not assignable to parameter of type 'number'.[39m

[31mâœ– [90mtest/[31ma.errors.ts[90m[31m:4:10:[39m[1m Wrong error
[22m  Expected: [32mArgument of type '"1"' is not assignable to parameter of type 'number'.[39m
  Got: [31mArgument of type 'string' is not assignable to parameter of type 'number'.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:9:7:[39m[1m Wrong error
[22m  Expected: [32mWrong[39m
  Got: [31mExpected 3 arguments, but got 2.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:12:20:[39m[1m Unexpected error
[22m  [31mExpected 1 arguments, but got 2.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:15:12:[39m[1m Wrong error
[22m  Expected: [32mnot assignable to parameter of type '"set" | "tick"'[39m
  Got: [31mArgument of type '"unknown"' is not assignable to parameter of type 'keyof Events'.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:11:1:[39m[1m Error was not found
[22m  [31mMissed[39m

"
`;

exports[`checks mixed tests 1`] = `
"[33m-[39m Check types
[32mâœ”[39m Check types
[32mâœ” [39m[90ma/[37merrors.ts[90m[39m
[32mâœ” [39m[90ma/[37mtypes.ts[90m[39m
"
`;

exports[`checks negative tests 1`] = `
"[33m-[39m Check types
[31mâœ–[39m Check types

[31mâœ– [90m[31ma.d.ts[90m[31m:1:18:[39m[1m Type error [90mTS7010[39m
[22m  [31m'a', which lacks return-type annotation, implicitly has an 'any' return type.[39m

[31mâœ– [90mtest/[31ma.errors.ts[90m[31m:4:10:[39m[1m Wrong error
[22m  Expected: [32mArgument of type '"1"' is not assignable to parameter of type 'number'.[39m
  Got: [31mArgument of type 'string' is not assignable to parameter of type 'number'.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:9:7:[39m[1m Wrong error
[22m  Expected: [32mWrong[39m
  Got: [31mExpected 3 arguments, but got 2.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:12:20:[39m[1m Unexpected error
[22m  [31mExpected 1 arguments, but got 2.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:15:12:[39m[1m Wrong error
[22m  Expected: [32mnot assignable to parameter of type '"set" | "tick"'[39m
  Got: [31mArgument of type '"unknown"' is not assignable to parameter of type 'keyof Events'.[39m

[31mâœ– [90mtest/[31mb.errors.ts[90m[31m:11:1:[39m[1m Error was not found
[22m  [31mMissed[39m

"
`;

exports[`checks positive tests 1`] = `
"[33m-[39m Check types
[31mâœ–[39m Check types

[31mâœ– [90mtest/[31mtypes.ts[90m[31m:3:14:[39m[1m Type error [90mTS2345[39m
[22m  [31mArgument of type 'string' is not assignable to parameter of type 'number'.[39m

[31mâœ– [90mtest/[31merrors.ts[90m[31m:4:14:[39m[1m Wrong error
[22m  Expected: [32mArgument of type '"1"' is not assignable to parameter of type 'number'.[39m
  Got: [31mArgument of type 'string' is not assignable to parameter of type 'number'.[39m

"
`;

exports[`loads custom tsconfig.json 1`] = `
"[33m-[39m Check types
[32mâœ”[39m Check types
[32mâœ” [39m[90m[37mtypes.ts[90m[39m
"
`;

exports[`loads custom tsconfig.json with extends property 1`] = `
"[33m-[39m Check types
[32mâœ”[39m Check types
[32mâœ” [39m[90m[37mindex.errors.ts[90m[39m
"
`;

exports[`shows errors not associated with a file 1`] = `
"[33m-[39m Check types
[31mâœ–[39m Check types

undefined

"
`;

exports[`supports simple cases 1`] = `
"[33m-[39m Check types
[32mâœ”[39m Check types
[32mâœ” [39m[90mtest/[37merrors.ts[90m[39m
[32mâœ” [39m[90mtest/[37mtypes.ts[90m[39m
"
`;
